ykbh/src/lib.rs

```diff
@@ -58,34 +80,62 @@ impl SIRInterpreter {
-            match stmt {
-                Statement::MkRef(dest, src) => self.mkref(dest, src),
-                Statement::DynOffs { .. } => todo!(),
-                Statement::Store(dest, src) => self.store(dest, src),
-                Statement::BinaryOp { .. } => todo!(),
-                Statement::Nop => {}
-                Statement::Unimplemented(_) | Statement::Debug(_) => todo!(),
-                Statement::Cast(..) => todo!(),
-                Statement::Call(..) | Statement::StorageDead(_) => unreachable!(),
+        loop {
+            for stmt in body.blocks[self.bbidx].stmts.iter() {
+                match stmt {
+                    Statement::MkRef(dest, src) => self.mkref(dest, src),
+                    Statement::DynOffs { .. } => todo!(),
+                    Statement::Store(dest, src) => self.store(dest, src),
+                    Statement::BinaryOp { .. } => todo!(),
+                    Statement::Nop => {}
+                    Statement::Unimplemented(_) | Statement::Debug(_) => todo!(),
+                    Statement::Cast(..) => todo!(),
+                    Statement::Call(..) | Statement::StorageDead(_) => unreachable!(),
+                }
             }
-        }
 
-        match &body.blocks[0].term {
-            Terminator::Call {
-                operand: _op,
-                args: _args,
-                destination: _dest,
-            } => todo!(),
-            Terminator::Return => {}
-            t => todo!("{}", t),
+            match &body.blocks[self.bbidx].term {
+                Terminator::Call {
+                    operand: op,
+                    args: _args,
+                    destination: dest,
+                } => {
+                    let fname = if let CallOperand::Fn(sym) = op {
+                        sym
+                    } else {
+                        todo!("unknown call target");
+                    };
+
+                    // Initialise the new stack frame.
+                    let body = SIR.body(fname).unwrap();
+                    let frame = SIRInterpreter::allocate_locals(&body.local_decls);
+                    self.frames.push(frame);
+                    self.bbidx = 0;
+
+                    self.interpret(&body);
+                    // Get pointer to result from current frame.
+                    let ptr = self.local_ptr(&Local(0));
+                    // Restore previous stack frame, but keep the other frame around so the pointer to
+                    // the return value stays valid until we've copied it.
+                    let mut oldframe = self.frames.pop().unwrap();
+                    // Write results to destination.
+                    if let Some((dest, bbidx)) = dest {
+                        self.write(dest, ptr);
+                        self.bbidx = usize::try_from(*bbidx).unwrap();
+                    }
+                    // We've copied the value, so the old frame can be freed.
+                    oldframe.free();
```

Comment:
Could frame have a `Drop` so we don't have to do this explicitly?

Reply:
I thought about this too, but wasn't sure whether Rust would drop `oldframe` before we've written the result to the destination. So I went the safe route.

Reply:
I missed this: manual `free` is an accident waiting to happen. Rust has very predictable (and documented!) semantics for when things are dropped: assuming that works here (which I suspect it does), a `Drop` is far preferable.

Reply:
Okay, I didn't know the semantics were predictable. I assumed Rust would just drop the value anytime after it's not used anymore, which could be immediately after its last use (which would break this code), or at the end of the block (which would be fine). Let me double check the semantics and then I'll fix it.

Reply:
I've read up on drops, and you're right that `oldframe` won't be dropped until we've reached the end of the scope. So this is fixed in 0f7c951.